### PR TITLE
Distinct Query Continuation Token Support

### PR DESCRIPTION
**Pull request by bchong95**
# Distinct Query Continuation Token Support

## Description

Adding continuation token support for DISTINCT queries. The general idea is that users currently aren't able to run DISTINCT queries through the continuation token API without a matching ORDER BY clause. It's very limiting for these users. The solution was to add continuation token support for these types of queries. The general continuation token is like so:

```JSON
{
    "SourceContinuationToken": <SourceContinuationToken>,
    "DistinctMapContinuationToken": <DistinctMapContinuationToken>
}
```

This means we are serializing both the base continuation token and serializing the distinct map. If it's a DISTINCT query + matching ORDER BY, then the token is just the most recent hash. Otherwise we dump the whole hash table which looks like this:

```JSON
{
    "Numbers": [<Number64>, <Number64>, ..., <Number64>],
    "StringsLength4": [<uint32>, <uint32>, ..., <uint32>],
    "StringsLength8": [<ulong>, <ulong>, ..., <ulong>],
     .
     .
     .
     "StringsLength24Plus": [<UInt192>, <UInt192>, ..., <UInt192>],
     "Arrays": [<UInt192>, <UInt192>, ..., <UInt192>],
     "Object": [<UInt192>, <UInt192>, ..., <UInt192>],
}
```

Basically it's a dictionary of hashes which represents the state of the operator.

This dictionary is serialized using binary encoding, since the hashes are binary. The whole byte array is then Base64 encoded, so that it can go in a continuation token.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by kirankumarkolli** [comment]
Will CTLimitsinKB impact the completeness of CT for distinct's?

File: Microsoft.Azure.Cosmos/src/CosmosElements/Numbers/CosmosFloat32.cs [hunk 1]
```diff
@@ -57,7 +57,7 @@ public override void WriteTo(IJsonWriter jsonWriter)
             jsonWriter.WriteFloat32Value(this.GetValue());
         }
 
-        protected abstract float GetValue();
+        public abstract float GetValue();
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosElements/Numbers/CosmosFloat32.cs]: >public [](start = 8, length = 6)

Can it be internal?

**Comment by kirankumarkolli** [comment]
Lets please sync-up once what it means for other SDK's? 
Replicating all of these including CosmosElements is a very expensive work.

File: Microsoft.Azure.Cosmos/src/CosmosElements/Numbers/CosmosFloat32.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/CosmosElements/Numbers/CosmosFloat32.cs]: The class is already internal.

**Comment by bchong95** [comment]
There is no replication for other SDKs. As far as I know the other SDKs gave up on supporting continuation tokens a long time ago.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/DistinctDocumentQueryExecutionComponent.cs [hunk 2]
```diff
@@ -126,31 +115,49 @@ public override async Task<QueryResponseCore> DrainAsync(int maxElements, Cancel
...
+                diagnostics: sourceResponse.diagnostics,
+                responseLengthBytes: sourceResponse.ResponseLengthBytes);
         }
 
-        public override bool TryGetContinuationToken(out string state)
+        public override bool TryGetContinuationToken(out string continuationToken)
         {
-            state = null;
-            return false;
+            if (!this.IsDone)
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/DistinctDocumentQueryExecutionComponent.cs]: Once aggregates get merged in only to this in the compute fork.

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctMap.UnorderedDistinctMap.cs [hunk 3]
```diff
@@ -303,6 +462,162 @@ private bool AddObjectValue(CosmosObject cosmosObject)
...
                 return this.objects.Add(hash);
             }
+
+            public static UnorderdDistinctMap Create(string continuationToken)
+            {
+                HashSet<Number64> numbers = new HashSet<Number64>();
+                HashSet<uint> stringsLength4 = new HashSet<uint>();
+                HashSet<ulong> stringsLength8 = new HashSet<ulong>();
+                HashSet<UInt128> stringsLength16 = new HashSet<UInt128>();
+                HashSet<UInt192> stringsLength24 = new HashSet<UInt192>();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/DistinctMap.UnorderedDistinctMap.cs]: >stringsLength24 [](start = 33, length = 15)

cap it at 16 bytes

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctMap.UnorderedDistinctMap.cs [hunk 4]
```diff
@@ -303,6 +462,162 @@ private bool AddObjectValue(CosmosObject cosmosObject)
...
+                HashSet<UInt192> stringsLength24Plus = new HashSet<UInt192>();
+                HashSet<UInt192> arrays = new HashSet<UInt192>();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/DistinctMap.UnorderedDistinctMap.cs]: >UInt192 [](start = 24, length = 7)

128

File: Microsoft.Azure.Cosmos/src/Query/Core/UInt192.cs [hunk 5]
```diff
@@ -440,5 +443,47 @@ public ulong GetLow()
...
+            /// <param name="writer">The writer.</param>
+            /// <param name="value">The value.</param>
+            /// <param name="serializer">The serializer.</param>
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                UInt192 uint192 = (UInt192)value;
+                JToken token = JToken.FromObject(uint192.ToString());
+                token.WriteTo(writer);
+            }
+        }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/UInt192.cs]: double check if we need this

**Comment by bchong95** [comment]
only shipping through TryGetContinuationToken.

---
In reply to: [535514974](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/851#issuecomment-535514974) [](ancestors = 535514974)
